File: internal/ring/ring.go
// Package ring defines the ordered deployment rings that every application is
// promoted through. Rings are shared across all applications.
//
// ADDING A RING is a single-place change: append an entry to the `ordered`
// slice below (in the correct position). Everything else — promotion order,
// API responses, and the UI — is derived from this list.
package ring

// Ring is one stage in the shared promotion pipeline.
type Ring struct {
	// Name is the stable identifier used in the API, config and storage.
	Name string `json:"name"`
	// Label is a human-friendly description shown in the UI.
	Label string `json:"label"`
}

// ordered is the single source of truth for the ring pipeline, lowest
// environment first. Promotion always moves from index i to index i+1.
var ordered = []Ring{
	{Name: "ring0", Label: "Dev"},
	{Name: "ring1", Label: "Integration"},
	{Name: "ring2", Label: "Acceptance"},
	{Name: "ring3", Label: "Production"},
}

// All returns a copy of the ordered ring list.
func All() []Ring {
	out := make([]Ring, len(ordered))
	copy(out, ordered)
	return out
}

// Names returns the ordered ring names.
func Names() []string {
	out := make([]string, len(ordered))
	for i, r := range ordered {
		out[i] = r.Name
	}
	return out
}

// Index returns the position of the named ring, or -1 if it does not exist.
func Index(name string) int {
	for i, r := range ordered {
		if r.Name == name {
			return i
		}
	}
	return -1
}

// Get returns the ring with the given name.
func Get(name string) (Ring, bool) {
	if i := Index(name); i >= 0 {
		return ordered[i], true
	}
	return Ring{}, false
}

// IsValid reports whether name refers to a known ring.
func IsValid(name string) bool {
	return Index(name) >= 0
}

// Next returns the ring immediately after name in the pipeline. ok is false
// when name is the last ring (production) or is unknown.
func Next(name string) (Ring, bool) {
	i := Index(name)
	if i < 0 || i+1 >= len(ordered) {
		return Ring{}, false
	}
	return ordered[i+1], true
}

File: ios/RingPromoter/Models/Ring.swift
import Foundation

/// One stage in the shared promotion pipeline.
///
/// Mirrors `ring.Ring` in the backend. The ordered list of rings is served by
/// `GET /api/apps` — the app never hard-codes it, so adding a ring server-side
/// flows through without an app release.
struct Ring: Codable, Hashable, Sendable, Identifiable {
    /// Stable identifier used in every API path and body.
    let name: String
    /// Human-friendly label shown in the UI ("Integration", "Production").
    let label: String

    var id: String { name }

    enum CodingKeys: String, CodingKey {
        case name, label
    }
}

/// The ordered ring pipeline for this control plane.
///
/// Wraps the server's ring list so promotion legality ("what is the next ring?",
/// "is this the last ring?") is answered in one place instead of being
/// re-derived at each call site.
struct RingPipeline: Hashable, Sendable {
    let rings: [Ring]

    init(_ rings: [Ring]) {
        self.rings = rings
    }

    /// The last ring — the one a production password protects.
    var productionRing: Ring? { rings.last }

    var names: [String] { rings.map(\.name) }

    func index(of name: String) -> Int? {
        rings.firstIndex { $0.name == name }
    }

    func ring(named name: String) -> Ring? {
        rings.first { $0.name == name }
    }

    /// The ring immediately after `name`, or nil when `name` is the last ring
    /// or is unknown. Promotion always targets exactly this ring — the app must
    /// never offer a target the server would refuse.
    func next(after name: String) -> Ring? {
        guard let i = index(of: name), i + 1 < rings.count else { return nil }
        return rings[i + 1]
    }

    /// Whether promoting out of `name` would deploy into the production ring.
    /// This is the condition that requires the production password.
    func promotingTargetsProduction(from name: String) -> Bool {
        guard let next = next(after: name), let prod = productionRing else { return false }
        return next.name == prod.name
    }

    /// Whether seeding `name` directly deploys into the production ring.
    func isProduction(_ name: String) -> Bool {
        productionRing?.name == name
    }

    static let empty = RingPipeline([])
}
